<h1 align=center><font size = 5>Week 3 Assignment : Segmenting and Clustering Neighborhoods in Toronto</font></h1>

# Task One: Download Toronto Dataset

For this assignment, I will be required to explore and cluster the neighborhoods in Toronto. This is the Jupyter notebook created for this interesting assignment. First of all, we will use *pandas.read_html(url)* function to scrape the neighborhood information from this Wikipedia web page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.

Turn out, it is relatively straightforward the information from the web page. *pandas.read_html(url)* requires one additional tool to be installed in the conda which is *lxml*.


For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

1. Start by creating a new Notebook for this assignment.

2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

3. To create the above dataframe:

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

4. Submit a link to your Notebook on your Github repository. (10 marks)

Note: There are different website scraping libraries and packages in Python. For scraping the above table, you can simply use pandas to read the table into a pandas dataframe.

Another way, which would help to learn for more complicated cases of web scraping is using the BeautifulSoup package. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

In [1]:
import pandas as pd

Install the *lxml* tool in the Jupyter notebook environment.

In [2]:
!conda install -c anaconda lxml --yes

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - lxml


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.1.1   |                0         132 KB  anaconda
    certifi-2020.4.5.1         |           py36_0         159 KB  anaconda
    libxslt-1.1.33             |       h7d1a2b0_0         577 KB  anaconda
    lxml-4.5.0                 |   py36hefd8a0e_0         1.6 MB  anaconda
    openssl-1.1.1g             |       h7b6447c_0         3.8 MB  anaconda
    ------------------------------------------------------------
                                           Total:         6.2 MB

The following NEW packages will be INSTALLED:

  libxslt            anaconda/linux-64::libxslt-1.1.33-h7d1a2b0_0
  lxml               anaconda/linux-64::lxml-4.5.0-py36hefd8a0e_0

The following packages wi

Read the Wikipedia web page using *pandas.read_html(url)*

In [3]:
dfs = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

In [4]:
dfs[0].head(10)

Postal Code           Borough                                 Neighborhood
0         M1A      Not assigned                                          NaN
1         M2A      Not assigned                                          NaN
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                          NaN
8         M9A         Etobicoke                             Islington Avenue
9         M1B       Scarborough                               Malvern, Rouge

Let's display the *shape* of the data.

In [5]:
dfs[0].shape

(180, 3)

Now comes the *wrangle* the data and data *cleaning* part. Let's query all the data which *Borough* is not *Not assigned*.

In [6]:
toronto_df = dfs[0].loc[dfs[0]['Borough'] != 'Not assigned']

In [7]:
toronto_df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168              Business reply mail Processing Centre  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking make sure that *Not assigned* value is not in *Neighborhood* field.

In [8]:
toronto_df.loc[toronto_df['Neighborhood'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

In [9]:
toronto_df.loc[toronto_df['Neighborhood'] == 'NaN']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

Checking the *Borough* values.

In [10]:
toronto_df.groupby('Borough')['Postal Code'].nunique()

Borough
Central Toronto      9
Downtown Toronto    19
East Toronto         5
East York            5
Etobicoke           12
Mississauga          1
North York          24
Scarborough         17
West Toronto         6
York                 5
Name: Postal Code, dtype: int64

Checking the *Postal Code*

In [11]:
groupby_df = toronto_df.groupby('Postal Code')['Postal Code'].nunique()

Checking the *groupby_df.shape*

In [12]:
groupby_df.shape

(103,)

Finally, checking the *toronto_df.shape*

In [13]:
toronto_df.shape

(103, 3)

# Task Two: Integrate the Latitude and Longitude Coordinates Dataset for Toronto Dataframe

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Taking postal code M5G as an example, your code would look something like this:
    
Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

Use the Geocoder package or the csv file to create the following dataframe:

### Option 1

Download CSV and join both DataFrames together

In [14]:
!wget -q -O 'geospatial_data.csv' http://cocl.us/Geospatial_data
print('Geospatial data downloaded!')

Geospatial data downloaded!


In [15]:
geospatial_df = pd.read_csv('geospatial_data.csv')
geospatial_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
toronto_geospatial_df = pd.merge(toronto_df, geospatial_df, left_on='Postal Code', right_on='Postal Code', how='left')

In [17]:
toronto_geospatial_df.loc[toronto_geospatial_df['Postal Code'] == 'M5G']

Postal Code           Borough        Neighborhood   Latitude  Longitude
24         M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383

In [18]:
toronto_geospatial_df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100              Business reply mail Processing Centre  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

### Option 2

Query the GPS data using *geocoder*.

In [19]:
#!conda install -c anaconda geocoder --yes
!conda install -c conda-forge geocoder --yes

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    brotlipy-0.7.0             |py36h8c4c3a4_1000         346 KB  conda-forge
    chardet-3.0.4              |py36h9f0ad1d_1006         188 KB  conda-forge
    click-7.1.2                |     pyh9f0ad1d_0          64 KB  conda-forge
    cryptography-2.9.2         |   py36h45558ae_0         613 KB  conda-forge
    future-0.18.2              |   py36h9f0ad1d_1         714 KB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    pysocks-1.7.1              |   py36h9f0ad1d_1          27 KB  conda-forge
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    ------------------------------------------------------------
                  

In [20]:
import geocoder

for index, data in toronto_geospatial_df.iterrows(): 
    #print(index, data['Postal Code'])
    postal_code = data['Postal Code']
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    #while(lat_lng_coords is None):
      #g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
      #lat_lng_coords = g.latlng
    #latitude = lat_lng_coords[0]
    #longitude = lat_lng_coords[1]
    #print(latitude)
    #print(longitude)

Output the DataFrame after Latitude and Longitude have been integrated.

In [21]:
toronto_geospatial_df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100              Business reply mail Processing Centre  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

# Task Three: Explore and Cluster the Neighborhoods in Toronto

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.

Just make sure:

to add enough Markdown cells to explain what you decided to do and to report any observations you make.
to generate maps to visualize your neighborhoods and how they cluster together.
Once you are happy with your analysis, submit a link to the new Notebook on your Github repository.

In [22]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-1.22.0-pyh9f0ad1d_0



geopy-1.22.0         | 63 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: 

In [23]:
# instantiate the dataframe
neighborhoods = toronto_geospatial_df

In [24]:
neighborhoods

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

And make sure that the dataset for **Toronto, Ontario (Canada)** has all 10 boroughs and 103 neighborhoods.

In [25]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


### Use geopy library to get the latitude and longitude values of Toronto, Ontario (Canada).

In order to define an instance of the **geocoder**, we need to define a user_agent. We will name our agent *toronto_explorer*, as shown below.

In [27]:
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Ontario (Canada) are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Ontario (Canada) are 43.6534817, -79.3839347.


### Create a map of Toronto, Ontario (Canada) with neighborhoods superimposed on top.

In [32]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

**Folium** is a great visualization library. Let's zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in **Downtown Toronto**. So let's slice the original dataframe and create a new dataframe of the **Downtown Toronto** data.

In [29]:
downtown_toronto_data = neighborhoods[neighborhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_toronto_data.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Also take a look at neighborhood for **Central Toronto**. So let's slice the original dataframe and create a new dataframe of the **Central Toronto** data.

In [30]:
central_toronto_data = neighborhoods[neighborhoods['Borough'] == 'Central Toronto'].reset_index(drop=True)
central_toronto_data.head()

Postal Code          Borough              Neighborhood   Latitude  Longitude
0         M4N  Central Toronto             Lawrence Park  43.728020 -79.388790
1         M5N  Central Toronto                  Roselawn  43.711695 -79.416936
2         M4P  Central Toronto          Davisville North  43.712751 -79.390197
3         M5P  Central Toronto  Forest Hill North & West  43.696948 -79.411307
4         M4R  Central Toronto        North Toronto West  43.715383 -79.405678

Plus also take a look at neighborhood for **North York**. So let's slice the original dataframe and create a new dataframe of the **North York** data.

In [31]:
north_york_data = neighborhoods[neighborhoods['Borough'] == 'North York'].reset_index(drop=True)
north_york_data.head()

Postal Code     Borough                      Neighborhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.464763  
3 -79.352188  
4 -79.445073

### Define Foursquare Credentials and Version

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [93]:
CLIENT_ID = 'XXXXXXXX' # your Foursquare ID
CLIENT_SECRET = 'XXXXXXXXXX' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XXXXXXXX
CLIENT_SECRET:XXXXXXXXXX


Let's explore the first neighborhood in our dataframe - **north_york_data**.

Get the neighborhood's name.

In [35]:
north_york_data.loc[0, 'Neighborhood']

'Parkwoods'

Get the neighborhood's latitude and longitude values.

In [36]:
neighborhood_latitude = north_york_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = north_york_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = north_york_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Now, let's get the top 100 venues that are in **Parkwoods** within a radius of *500* meters.

First, let's create the GET request URL. Name your URL url.

In [94]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=XXXXXXXX&client_secret=XXXXXXXXXX&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

Send the GET request and examine the resutls

In [38]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ec96852b9a389001bfecca2'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [39]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [40]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

And how many venues were returned by Foursquare?

In [41]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare.


### Explore Neighborhoods in North York, Toronto

Let's create a function to repeat the same process to all the neighborhoods in North York, Toronto

In [42]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Use above function on each neighborhood and create a new dataframe called north_york_venues.

In [43]:
north_york_venues = getNearbyVenues(names=north_york_data['Neighborhood'],
                                   latitudes=north_york_data['Latitude'],
                                   longitudes=north_york_data['Longitude']
                                  )

Parkwoods
Victoria Village
Lawrence Manor, Lawrence Heights
Don Mills
Glencairn
Don Mills
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Fairview, Henry Farm, Oriole
Northwood Park, York University
Bayview Village
Downsview
York Mills, Silver Hills
Downsview
North Park, Maple Leaf Park, Upwood Park
Humber Summit
Willowdale, Newtonbrook
Downsview
Bedford Park, Lawrence Manor East
Humberlea, Emery
Willowdale
Downsview
York Mills West
Willowdale


Take a quick look at **north_york_venues**

In [47]:
north_york_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

Let's check the size of the resulting dataframe

In [44]:
print(north_york_data.shape)
north_york_data.head()

(24, 5)


Postal Code     Borough                      Neighborhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.464763  
3 -79.352188  
4 -79.445073

Let's check how many venues were returned for each neighborhood

In [45]:
north_york_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   24   
Don Mills                                                           27   
Downsview                                                           19   
Fairview, Henry Farm, Oriole                                        67   
Glencairn                                                            4   
Hillcrest Village                                                    4   
Humber Summit                                                        2   
Humberlea, Emery                                                     1   
Lawrence Manor, Lawrence Heights                                    13   
North Park, Maple Leaf Park, Upwood Park                             4   
Northwood Park, York University                                      7   
Parkwoods                                                            2   
Victoria Village                                                     5   
Willowdale                                                          37   
York Mills West                                                      4   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    24   
Don Mills                                                            27   
Downsview                                                            19   
Fairview, Henry Farm, Oriole                                         67   
Glencairn                                                             4   
Hillcrest Village                                                     4   
Humber Summit                                                         2   
Humberlea, Emery                                                      1   
Lawrence Manor, Lawrence Heights                                     13   
North Park, Maple Leaf Park, Upwood Park                              4   
Northwood Park, York University                                       7   
Parkwoods                                                             2   
Victoria Village                                                      5   
Willowdale                                                           37   
York Mills West                                                       4   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   24              24   
Don Mills                                           27              27   
Downsview                                           19              19   
Fairview, Henry Farm, Oriole                        67              67   
Glencairn                                            4               4   
Hillcrest Village                                    4               4   
Humber Summit                                        2               2   
Humberlea, Emery                                     1               1   
Lawrence Manor, Lawrence Heights                    13              13   
North Park, Maple Leaf Park, Upwood Park             4               4   
Northwood Park, York University                      7               7   
Parkwoods                                            2               2   
Victoria Village                

Let's find out how many unique categories can be curated from all the returned venues

In [46]:
print('There are {} uniques categories.'.format(len(north_york_venues['Venue Category'].unique())))

There are 106 uniques categories.


### Analyze Each Neighborhood

Now come the interesting part, how can we analyze each neighborhood based the information we have gathered so far. Using the **one hot** technique will be the best option to *flip* the *Venue Category* into column view.

In [48]:
# one hot encoding
north_york_onehot = pd.get_dummies(north_york_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
north_york_onehot['Neighborhood'] = north_york_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [north_york_onehot.columns[-1]] + list(north_york_onehot.columns[:-1])
north_york_onehot = north_york_onehot[fixed_columns]

north_york_onehot.head()

Neighborhood  Accessories Store  Airport  American Restaurant  \
0         Parkwoods                  0        0                    0   
1         Parkwoods                  0        0                    0   
2  Victoria Village                  0        0                    0   
3  Victoria Village                  0        0                    0   
4  Victoria Village                  0        0                    0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Bakery  Bank  \
0                    0                 0                   0       0     0   
1                    0                 0                   0       0     0   
2                    0                 0                   0       0     0   
3                    0                 0                   0       0     0   
4                    0                 0                   0       0     0   

   Bar  Baseball Field  Basketball Court  Beer Store  Bike Shop  Boutique  \
0    0               0                 0           0          0         0   
1    0               0                 0           0          0         0   
2    0               0                 0           0          0         0   
3    0               0                 0           0          0         0   
4    0               0                 0           0          0         0   

   Breakfast Spot  Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  \
0               0            0                0             0              0   
1               0            0                0             0              0   
2               0            0                0             0              0   
3               0            0                0             0              0   
4               0            0                0             0              0   

   Bus Station  Bus Stop  Business Service  Butcher  Café  \
0            0         0                 0        0     0   
1            0         0                 0        0     0   
2            0         0                 0        0     0   
3            0         0                 0        0     0   
4            0         0                 0        0     0   

   Caribbean Restaurant  Chinese Restaurant  Chocolate Shop  Clothing Store  \
0                     0                   0               0               0   
1                     0                   0               0               0   
2                     0                   0               0               0   
3                     0                   0               0               0   
4                     0                   0               0               0   

   Coffee Shop  Comfort Food Restaurant  Concert Hall  \
0            0                        0             0   
1            0                        0             0   
2            0                        0             0   
3            1                        0             0   
4            0                        0             0   

   Construction & Landscaping  Convenience Store  Cosmetics Shop  \
0                           0                  0               0   
1                           0                  0               0   
2                           0                  0               0   
3                           0                  0               0   
4                           0                  0               0   

   Deli / Bodega  Department Store  Dessert Shop  Dim Sum Restaurant  Diner  \
0              0                 0             0                   0      0   
1              0                 0             0                   0      0   
2              0                 0             0                   0      0   
3              0                 0             0                   0      0   
4              0                 0             0                   0      0   

   Discount Store  Dog Run  Electronics Store  Event Space  \
0               0        0    

And let's examine the new dataframe size.

In [49]:
north_york_onehot.shape

(245, 107)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [50]:
north_york_grouped = north_york_onehot.groupby('Neighborhood').mean().reset_index()
north_york_grouped

Neighborhood  Accessories Store  \
0   Bathurst Manor, Wilson Heights, Downsview North           0.000000   
1                                   Bayview Village           0.000000   
2                 Bedford Park, Lawrence Manor East           0.000000   
3                                         Don Mills           0.000000   
4                                         Downsview           0.000000   
5                      Fairview, Henry Farm, Oriole           0.000000   
6                                         Glencairn           0.000000   
7                                 Hillcrest Village           0.000000   
8                                     Humber Summit           0.000000   
9                                  Humberlea, Emery           0.000000   
10                 Lawrence Manor, Lawrence Heights           0.076923   
11         North Park, Maple Leaf Park, Upwood Park           0.000000   
12                  Northwood Park, York University           0.000000   
13                                        Parkwoods           0.000000   
14                                 Victoria Village           0.000000   
15                                       Willowdale           0.000000   
16                                  York Mills West           0.000000   

     Airport  American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0   0.000000             0.000000             0.000000          0.000000   
1   0.000000             0.000000             0.000000          0.000000   
2   0.000000             0.041667             0.000000          0.000000   
3   0.000000             0.000000             0.000000          0.074074   
4   0.052632             0.000000             0.000000          0.000000   
5   0.000000             0.014925             0.000000          0.014925   
6   0.000000             0.000000             0.000000          0.000000   
7   0.000000             0.000000             0.000000          0.000000   
8   0.000000             0.000000             0.000000          0.000000   
9   0.000000             0.000000             0.000000          0.000000   
10  0.000000             0.000000             0.000000          0.000000   
11  0.000000             0.000000             0.000000          0.000000   
12  0.000000             0.000000             0.000000          0.000000   
13  0.000000             0.000000             0.000000          0.000000   
14  0.000000             0.000000             0.000000          0.000000   
15  0.000000             0.000000             0.027027          0.000000   
16  0.000000             0.000000             0.000000          0.000000   

    Athletics & Sports    Bakery      Bank       Bar  Baseball Field  \
0             0.000000  0.000000  0.095238  0.000000        0.000000   
1             0.000000  0.000000  0.250000  0.000000        0.000000   
2             0.000000  0.000000  0.000000  0.000000        0.000000   
3             0.037037  0.000000  0.000000  0.000000        0.000000   
4             0.052632  0.000000  0.052632  0.000000        0.052632   
5             0.000000  0.029851  0.029851  0.014925        0.014925   
6             0.000000  0.000000  0.000000  0.000000        0.000000   
7             0.000000  0.000000  0.000000  0.000000        0.000000   
8             0.000000  0.000000  0.000000  0.000000        0.000000   
9             0.000000  0.000000  0.000000  0.000000        1.000000   
10            0.000000  0.000000  0.000000  0.000000        0.000000   
11            0.000000  0.250000  0.000000  0.000000        0.000000   
12            0.000000  0.000000  0.000000  0.142857        0.000000   
13            0.000000  0.000000  0.000000  0.000000        0.000000   
14            0.000000  0.000000  0.000000  0.000000        0.000000   
15            0.000000  0.000000  0.027027  0.000000        0.000000   
16            0.000000  0.000000  0.250000  0.000000        0.000000   

    Basketball Court  Beer Store  Bike S

Let's confirm the new size

In [51]:
north_york_grouped.shape

(17, 107)

#### Let's print each neighborhood along with the top 5 most common venues

In [65]:
num_top_venues = 5

for hood in north_york_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = north_york_grouped[north_york_grouped['Neighborhood'] == hood].T.reset_index()
    #print('1. ', temp)
    temp.columns = ['venue','freq']
    #print('2. ', temp)
    temp = temp.iloc[1:]
    #print('3. ', temp)
    temp['freq'] = temp['freq'].astype(float)
    #print('4. ', temp)
    temp = temp.round({'freq': 2})
    #print(temp)
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0                Coffee Shop  0.10
1                       Bank  0.10
2                Supermarket  0.05
3  Middle Eastern Restaurant  0.05
4                       Park  0.05


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Café  0.25
2                 Bank  0.25
3  Japanese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0         Coffee Shop  0.08
1      Sandwich Place  0.08
2  Italian Restaurant  0.08
3         Pizza Place  0.04
4                Café  0.04


----Don Mills----
                 venue  freq
0                 Café  0.07
1           Restaurant  0.07
2     Asian Restaurant  0.07
3                  Gym  0.07
4  Japanese Restaurant  0.07


----Downsview----
              venue  freq
0     Grocery Store  0.16
1              Park  0.11
2  Business Service  0.11
3   

Let's put that into a pandas dataframe

First, let's write a function to sort the venues in descending order.

In [66]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [68]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = north_york_grouped['Neighborhood']

for ind in np.arange(north_york_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(north_york_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0   Bathurst Manor, Wilson Heights, Downsview North                  Bank   
1                                   Bayview Village    Chinese Restaurant   
2                 Bedford Park, Lawrence Manor East    Italian Restaurant   
3                                         Don Mills                  Café   
4                                         Downsview         Grocery Store   
5                      Fairview, Henry Farm, Oriole        Clothing Store   
6                                         Glencairn                  Park   
7                                 Hillcrest Village           Golf Course   
8                                     Humber Summit           Pizza Place   
9                                  Humberlea, Emery        Baseball Field   
10                 Lawrence Manor, Lawrence Heights        Clothing Store   
11         North Park, Maple Leaf Park, Upwood Park                  Park   
12                  Northwood Park, York University    Miscellaneous Shop   
13                                        Parkwoods     Food & Drink Shop   
14                                 Victoria Village           Coffee Shop   
15                                       Willowdale           Pizza Place   
16                                  York Mills West                  Park   

         2nd Most Common Venue       3rd Most Common Venue  \
0                  Coffee Shop                 Pizza Place   
1                         Bank                        Café   
2                  Coffee Shop              Sandwich Place   
3                          Gym         Japanese Restaurant   
4             Business Service                        Park   
5                  Coffee Shop        Fast Food Restaurant   
6          Japanese Restaurant            Sushi Restaurant   
7     Mediterranean Restaurant                        Pool   
8                          Gym               Women's Store   
9                Women's Store              Discount Store   
10               Women's Store          Miscellaneous Shop   
11  Construction & Landscaping                      Bakery   
12                 Coffee Shop        Caribbean Restaurant   
13                        Park               Women's Store   
14                 Pizza Place  Financial or Legal Service   
15                 Coffee Shop            Ramen Restaurant   
16           Convenience Store                        Bank   

   4th Most Common Venue    5th Most Common Venue    6th Most Common Venue  \
0             Restaurant       Chinese Restaurant           Sandwich Place   
1    Japanese Restaurant               Food Court       Dim Sum Restaurant   
2             Restaurant           Breakfast Spot                      Pub   
3             Restaurant               Beer Store              Coffee Shop   
4                  Hotel                 Bus Stop           Discount Store   
5             Restaurant      Japanese Restaurant               Shoe Store   
6                    Pub           Clothing Store              Coffee Shop   
7                Dog Run            Women's Store             Dessert Shop   
8                  Diner              Coffee Shop  Comfort Food Restaurant   
9            Coffee Shop  Comfort Food Restaurant             Concert Hall   
10              Boutique              Coffee Shop              Event Space   
11      Basketball Court                    Diner           Clothing Store   
12         Metro Station           Massage Studio       Falafel Restaurant   
13    Dim Sum Restaurant           Clothing Store              Coffee Shop   
14          Hockey Arena    Portuguese Restaurant            Women's Store   
15        Sandwich Place                     Café            Grocery Store   
16         Women's Store       Dim Sum Restaurant           Clothing Store   

         7th Most Common Venue       8th Most Common Venue  \
0               Ice Cream Shop               Shopping Mall   
1     

### Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [69]:
# set number of clusters
kclusters = 5

north_york_grouped_clustering = north_york_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(north_york_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 4, 3, 2], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [70]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

north_york_merged = north_york_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
north_york_merged = north_york_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

north_york_merged.head() # check the last columns!

Postal Code     Borough                      Neighborhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656             0.0     Food & Drink Shop                  Park   
1 -79.315572             1.0           Coffee Shop           Pizza Place   
2 -79.464763             1.0        Clothing Store         Women's Store   
3 -79.352188             1.0                  Café                   Gym   
4 -79.445073             1.0                  Park   Japanese Restaurant   

        3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0               Women's Store    Dim Sum Restaurant         Clothing Store   
1  Financial or Legal Service          Hockey Arena  Portuguese Restaurant   
2          Miscellaneous Shop              Boutique            Coffee Shop   
3         Japanese Restaurant            Restaurant             Beer Store   
4            Sushi Restaurant                   Pub         Clothing Store   

  6th Most Common Venue    7th Most Common Venue  8th Most Common Venue  \
0           Coffee Shop  Comfort Food Restaurant           Concert Hall   
1         Women's Store             Dessert Shop         Clothing Store   
2           Event Space   Furniture / Home Store  Vietnamese Restaurant   
3           Coffee Shop         Asian Restaurant     Athletics & Sports   
4           Coffee Shop  Comfort Food Restaurant           Concert Hall   

        9th Most Common Venue 10th Most Common Venue  
0  Construction & Landscaping      Convenience Store  
1     Comfort Food Restaurant           Concert Hall  
2           Accessories Store        Supplement Shop  
3              Sandwich Place     Italian Restaurant  
4  Construction & Landscaping      Convenience Store

In [74]:
north_york_merged.describe

<bound method NDFrame.describe of    Postal Code     Borough                                     Neighborhood  \
0          M3A  North York                                        Parkwoods   
1          M4A  North York                                 Victoria Village   
2          M6A  North York                 Lawrence Manor, Lawrence Heights   
3          M3B  North York                                        Don Mills   
4          M6B  North York                                        Glencairn   
5          M3C  North York                                        Don Mills   
6          M2H  North York                                Hillcrest Village   
7          M3H  North York  Bathurst Manor, Wilson Heights, Downsview North   
8          M2J  North York                     Fairview, Henry Farm, Oriole   
9          M3J  North York                  Northwood Park, York University   
10         M2K  North York                                  Bayview Village   
11         M3K  No

In [72]:
north_york_merged

Postal Code     Borough                                     Neighborhood  \
0          M3A  North York                                        Parkwoods   
1          M4A  North York                                 Victoria Village   
2          M6A  North York                 Lawrence Manor, Lawrence Heights   
3          M3B  North York                                        Don Mills   
4          M6B  North York                                        Glencairn   
5          M3C  North York                                        Don Mills   
6          M2H  North York                                Hillcrest Village   
7          M3H  North York  Bathurst Manor, Wilson Heights, Downsview North   
8          M2J  North York                     Fairview, Henry Farm, Oriole   
9          M3J  North York                  Northwood Park, York University   
10         M2K  North York                                  Bayview Village   
11         M3K  North York                                        Downsview   
12         M2L  North York                         York Mills, Silver Hills   
13         M3L  North York                                        Downsview   
14         M6L  North York         North Park, Maple Leaf Park, Upwood Park   
15         M9L  North York                                    Humber Summit   
16         M2M  North York                          Willowdale, Newtonbrook   
17         M3M  North York                                        Downsview   
18         M5M  North York                Bedford Park, Lawrence Manor East   
19         M9M  North York                                 Humberlea, Emery   
20         M2N  North York                                       Willowdale   
21         M3N  North York                                        Downsview   
22         M2P  North York                                  York Mills West   
23         M2R  North York                                       Willowdale   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   43.753259 -79.329656             0.0     Food & Drink Shop   
1   43.725882 -79.315572             1.0           Coffee Shop   
2   43.718518 -79.464763             1.0        Clothing Store   
3   43.745906 -79.352188             1.0                  Café   
4   43.709577 -79.445073             1.0                  Park   
5   43.725900 -79.340923             1.0                  Café   
6   43.803762 -79.363452             4.0           Golf Course   
7   43.754328 -79.442259             1.0                  Bank   
8   43.778517 -79.346556             1.0        Clothing Store   
9   43.767980 -79.487262             1.0    Miscellaneous Shop   
10  43.786947 -79.385975             1.0    Chinese Restaurant   
11  43.737473 -79.464763             1.0         Grocery Store   
12  43.757490 -79.374714             NaN                   NaN   
13  43.739015 -79.506944             1.0         Grocery Store   
14  43.713756 -79.490074             1.0                  Park   
15  43.756303 -79.565963             3.0           Pizza Place   
16  43.789053 -79.408493             NaN                   NaN   
17  43.728496 -79.495697             1.0         Grocery Store   
18  43.733283 -79.419750             1.0    Italian Restaurant   
19  43.724766 -79.532242             2.0        Baseball Field   
20  43.770120 -79.408493             1.0           Pizza Place   
21  43.761631 -79.520999             1.0         Grocery Store   
22  43.752758 -79.400049             0.0                  Park   
23  43.782736 -79.442259             1.0           Pizza Place   

         2nd Most Common Venue       3rd Most Common Venue  \
0                         Park               Women's Store   
1                  Pizza Place  Financial or Legal Service   
2                Women's Store          Miscellaneous Shop   
3                          Gym         Japanese Restaurant   
4          Japanese Restaurant            Sushi Restaurant   
5   

Watchout for the *NaN*

In [86]:
north_york_merged.dropna(subset = ["Cluster Labels"], inplace=True)
north_york_merged['Cluster Labels'] = north_york_merged['Cluster Labels'].astype(int)
north_york_merged

Postal Code     Borough                                     Neighborhood  \
0          M3A  North York                                        Parkwoods   
1          M4A  North York                                 Victoria Village   
2          M6A  North York                 Lawrence Manor, Lawrence Heights   
3          M3B  North York                                        Don Mills   
4          M6B  North York                                        Glencairn   
5          M3C  North York                                        Don Mills   
6          M2H  North York                                Hillcrest Village   
7          M3H  North York  Bathurst Manor, Wilson Heights, Downsview North   
8          M2J  North York                     Fairview, Henry Farm, Oriole   
9          M3J  North York                  Northwood Park, York University   
10         M2K  North York                                  Bayview Village   
11         M3K  North York                                        Downsview   
13         M3L  North York                                        Downsview   
14         M6L  North York         North Park, Maple Leaf Park, Upwood Park   
15         M9L  North York                                    Humber Summit   
17         M3M  North York                                        Downsview   
18         M5M  North York                Bedford Park, Lawrence Manor East   
19         M9M  North York                                 Humberlea, Emery   
20         M2N  North York                                       Willowdale   
21         M3N  North York                                        Downsview   
22         M2P  North York                                  York Mills West   
23         M2R  North York                                       Willowdale   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   43.753259 -79.329656               0     Food & Drink Shop   
1   43.725882 -79.315572               1           Coffee Shop   
2   43.718518 -79.464763               1        Clothing Store   
3   43.745906 -79.352188               1                  Café   
4   43.709577 -79.445073               1                  Park   
5   43.725900 -79.340923               1                  Café   
6   43.803762 -79.363452               4           Golf Course   
7   43.754328 -79.442259               1                  Bank   
8   43.778517 -79.346556               1        Clothing Store   
9   43.767980 -79.487262               1    Miscellaneous Shop   
10  43.786947 -79.385975               1    Chinese Restaurant   
11  43.737473 -79.464763               1         Grocery Store   
13  43.739015 -79.506944               1         Grocery Store   
14  43.713756 -79.490074               1                  Park   
15  43.756303 -79.565963               3           Pizza Place   
17  43.728496 -79.495697               1         Grocery Store   
18  43.733283 -79.419750               1    Italian Restaurant   
19  43.724766 -79.532242               2        Baseball Field   
20  43.770120 -79.408493               1           Pizza Place   
21  43.761631 -79.520999               1         Grocery Store   
22  43.752758 -79.400049               0                  Park   
23  43.782736 -79.442259               1           Pizza Place   

         2nd Most Common Venue       3rd Most Common Venue  \
0                         Park               Women's Store   
1                  Pizza Place  Financial or Legal Service   
2                Women's Store          Miscellaneous Shop   
3                          Gym         Japanese Restaurant   
4          Japanese Restaurant            Sushi Restaurant   
5                          Gym         Japanese Restaurant   
6     Mediterranean Restaurant                        Pool   
7                  Coffee Shop                 Pizza Place   
8                  Coffee Shop        Fast Food Restaurant   
9                  Coffee Shop        Caribbea

Finally, let's visualize the resulting clusters

In [83]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(north_york_merged['Latitude'], north_york_merged['Longitude'], north_york_merged['Neighborhood'], north_york_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 0

In [88]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 0, north_york_merged.columns[[2] + list(range(5, north_york_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
0         Parkwoods               0     Food & Drink Shop   
22  York Mills West               0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park         Women's Store    Dim Sum Restaurant   
22     Convenience Store                  Bank         Women's Store   

   5th Most Common Venue 6th Most Common Venue    7th Most Common Venue  \
0         Clothing Store           Coffee Shop  Comfort Food Restaurant   
22    Dim Sum Restaurant        Clothing Store              Coffee Shop   

      8th Most Common Venue       9th Most Common Venue  \
0              Concert Hall  Construction & Landscaping   
22  Comfort Food Restaurant                Concert Hall   

        10th Most Common Venue  
0            Convenience Store  
22  Construction & Landscaping

#### Cluster 1

In [89]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 1, north_york_merged.columns[[2] + list(range(5, north_york_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
1                                  Victoria Village               1   
2                  Lawrence Manor, Lawrence Heights               1   
3                                         Don Mills               1   
4                                         Glencairn               1   
5                                         Don Mills               1   
7   Bathurst Manor, Wilson Heights, Downsview North               1   
8                      Fairview, Henry Farm, Oriole               1   
9                   Northwood Park, York University               1   
10                                  Bayview Village               1   
11                                        Downsview               1   
13                                        Downsview               1   
14         North Park, Maple Leaf Park, Upwood Park               1   
17                                        Downsview               1   
18                Bedford Park, Lawrence Manor East               1   
20                                       Willowdale               1   
21                                        Downsview               1   
23                                       Willowdale               1   

   1st Most Common Venue       2nd Most Common Venue  \
1            Coffee Shop                 Pizza Place   
2         Clothing Store               Women's Store   
3                   Café                         Gym   
4                   Park         Japanese Restaurant   
5                   Café                         Gym   
7                   Bank                 Coffee Shop   
8         Clothing Store                 Coffee Shop   
9     Miscellaneous Shop                 Coffee Shop   
10    Chinese Restaurant                        Bank   
11         Grocery Store            Business Service   
13         Grocery Store            Business Service   
14                  Park  Construction & Landscaping   
17         Grocery Store            Business Service   
18    Italian Restaurant                 Coffee Shop   
20           Pizza Place                 Coffee Shop   
21         Grocery Store            Business Service   
23           Pizza Place                 Coffee Shop   

         3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
1   Financial or Legal Service          Hockey Arena  Portuguese Restaurant   
2           Miscellaneous Shop              Boutique            Coffee Shop   
3          Japanese Restaurant            Restaurant             Beer Store   
4             Sushi Restaurant                   Pub         Clothing Store   
5          Japanese Restaurant            Restaurant             Beer Store   
7                  Pizza Place            Restaurant     Chinese Restaurant   
8         Fast Food Restaurant            Restaurant    Japanese Restaurant   
9         Caribbean Restaurant         Metro Station         Massage Studio   
10                        Café   Japanese Restaurant             Food Court   
11                        Park                 Hotel               Bus Stop   
13                        Park                 Hotel               Bus Stop   
14                      Bakery      Basketball Court                  Diner   
17                        Park                 Hotel               Bus Stop   
18              Sandwich Place            Restaurant         Breakfast Spot   
20            Ramen Restaurant        Sandwich Place                   Café   
21                        Park                 Hotel               Bus Stop   
23            Ramen Restaurant        Sandwich Place                   Café   

   6th Most Common Venue    7th Most Common Venue    8th Most Common Venue  \
1          Women's Store             Dessert Shop           Clothing Store   
2            Event Space   Furniture / Home Store    Vietnamese Restaurant   
3            Coffee Shop         Asian Restaurant       Athletics & Sports   
4            Cof

#### Cluster 2

In [90]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 2, north_york_merged.columns[[2] + list(range(5, north_york_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
19  Humberlea, Emery               2        Baseball Field   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19         Women's Store        Discount Store           Coffee Shop   

      5th Most Common Venue 6th Most Common Venue       7th Most Common Venue  \
19  Comfort Food Restaurant          Concert Hall  Construction & Landscaping   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19     Convenience Store        Cosmetics Shop          Deli / Bodega

#### Cluster 3

In [91]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 3, north_york_merged.columns[[2] + list(range(5, north_york_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
15  Humber Summit               3           Pizza Place                   Gym   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
15         Women's Store                 Diner           Coffee Shop   

      6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
15  Comfort Food Restaurant          Concert Hall  Construction & Landscaping   

   9th Most Common Venue 10th Most Common Venue  
15     Convenience Store         Cosmetics Shop

#### Cluster 4

In [92]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 4, north_york_merged.columns[[2] + list(range(5, north_york_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
6  Hillcrest Village               4           Golf Course   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6  Mediterranean Restaurant                  Pool               Dog Run   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6         Women's Store          Dessert Shop           Coffee Shop   

     8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  
6  Comfort Food Restaurant          Concert Hall  Construction & Landscaping

That's it, finally coming to a *stop* for this assignment.